source/MaterialXRuntime/RtFileIo.cpp

```diff
@@ -621,46 +691,56 @@ namespace
 
         validateNodesHaveNodedefs(doc);
 
+        // Keep track of renamed nodes:
+        PvtRenamingMapper mapper;
+
         // Load all other elements.
         for (const ElementPtr& elem : doc->getChildren())
         {
             if (!filter || filter(elem))
             {
                 // Make sure the element has not been loaded already.
                 PvtPath path(ROOT_PATH + elem->getName());
-                if (stage->getPrimAtPath(path))
+                if (readOptions && readOptions->conflictResolution == RtReadOptions::SKIP_ELEMENTS && stage->getPrimAtPath(path))
                 {
                     continue;
                 }
 
                 if (elem->isA<Node>())
                 {
-                    readNode(elem->asA<Node>(), stage->getRootPrim(), stage);
+                    readNode(elem->asA<Node>(), stage->getRootPrim(), stage, mapper);
                 }
                 else if (elem->isA<NodeGraph>())
                 {
-                    readNodeGraph(elem->asA<NodeGraph>(), stage->getRootPrim(), stage);
+                    // Always skip if the nodegraph implements a nodedef
+                    if (readOptions && readOptions->conflictResolution == RtReadOptions::RENAME_ELEMENTS &&
+                        stage->getPrimAtPath(path) && elem->asA<NodeGraph>()->getNodeDef())
+                    {
+                        continue;
+                    }
+                    readNodeGraph(elem->asA<NodeGraph>(), stage->getRootPrim(), stage, mapper);
                 }
                 else
                 {
                     const RtToken category(elem->getCategory());
                     if (category != RtLook::typeName() &&
                         category != RtLookGroup::typeName() &&
                         category != RtMaterialAssign::typeName() &&
-                        category != RtCollection::typeName()) {
-                        readGenericPrim(elem, stage->getRootPrim(), stage);
+                        category != RtCollection::typeName() &&
+                        category != RtNodeDef::typeName()) {
```

Comment:
We were ending up with multiple copies of the nodedef, especially since they are first loaded in a separate loop at the top of the function.

Reply:
Yes nodedefs can't be resolved by renamed, they must be skipped.